pytensor/tensor/basic.py

```diff
@@ -3605,12 +3623,49 @@ def __setstate__(self, state):
             self.axis2 = 1
 
 
+def alloc_diag(diag, offset=0, axis1=0, axis2=1):
```

Comment:
Why the high level alloc diag is removed? Could it be Op from graph instead? Can it cause less efficient rewrites and more complicated graphs?

Reply:
The reason was twofold:
1. The JAX/Numba implementations were incorrect
2. It is relatively simple operation (set_subtensor and transpose) that should be efficient enough (and if anything benefit from rewrites we have in PyTensor, not be hurt by them). In the C-backend this should certainly be faster because it avoids the python switch overhead

Making it a symbolic Op solves both problems. We can consider making it an OpFromGraph subclass if we want to dispatch more specific implementations, but this shouldn't count as a regression because the pre-existing ones were simply wrong.

Reply:
They can see inside. They can also be inlined with `inline=True`. The dprint will move that complexity to inner graphs, whether that's easier or not I don't know.

There are however some issues with gradients of OpFromGraph #1, but we can definitely check if it works for this case. I am more inclined not to use it though. the graph isn't really that bad.